Proteomics/ProteolyticDigestion/PeptideWithSetModifications.cs

```diff
@@ -175,174 +175,336 @@ public string SequenceWithChemicalFormulas
-                case DissociationType.LowCID:
-                    skippers.Add((ProductType.b, 1));
-                    skippers.Add((ProductType.aDegree, 1));
-                    skippers.Add((ProductType.bDegree, 1));
-                    skippers.Add((ProductType.aStar, 1));
-                    skippers.Add((ProductType.bStar, 1));
-                    break;
-
-                case DissociationType.ETD:
-                case DissociationType.ECD:
-                case DissociationType.EThcD:
-                    skippers.AddRange(GetProlineZIonIndicies());
-                    break;
+                if (AllModsOneIsNterminus.TryGetValue(BaseSequence.Length + 2, out Modification mod))
+                {
+                    cTermMass += mod.MonoisotopicMass.Value;
+                }
             }
 
-            foreach (var productType in productCollection)
+            for (int r = 0; r < BaseSequence.Length - 1; r++)
             {
-                // we're separating the N and C terminal masses and computing a separate compact peptide for each one
-                // this speeds calculations up without producing unnecessary terminus fragment info
-                FragmentationTerminus temporaryFragmentationTerminus = TerminusSpecificProductTypes.ProductTypeToFragmentationTerminus[productType];
-                NeutralTerminusFragment[] terminalMasses = new CompactPeptide(this, temporaryFragmentationTerminus).TerminalMasses;
-
-                int firstUsableIndex = FullSequence.Length;
-                bool completeFragmentationSeries = true;
-                switch (productType)
+                // n-term fragments
+                if (calculateNTermFragments)
                 {
-                    case ProductType.aStar:
-                    case ProductType.bStar:
-                    case ProductType.yStar:
-                    case ProductType.aDegree:
-                    case ProductType.bDegree:
-                    case ProductType.yDegree:
+                    // get n-term residue mass
+                    if (Residue.TryGetResidue(BaseSequence[r], out Residue residue))
+                    {
+                        nTermMass += residue.MonoisotopicMass;
+                    }
+                    else
+                    {
+                        nTermMass = double.NaN;
+                    }
+
+                    // add side-chain mod
+                    if (AllModsOneIsNterminus.TryGetValue(r + 2, out Modification mod))
+                    {
+                        nTermMass += mod.MonoisotopicMass.Value;
+                    }
+
+                    // handle star and degree ions for low-res CID
+                    if (dissociationType == DissociationType.LowCID)
+                    {
+                        if (BaseSequence[r] == 'R' || BaseSequence[r] == 'K' || BaseSequence[r] == 'N' || BaseSequence[r] == 'Q')
```

Comment:
I understand we're in a "don't allocate memory" stance right now, but it was my understanding that if you have to look up an index (e.g. BaseSequence[r]) more than twice, then it was faster to store it (e.g. char currentResidue = BaseSequence[r]). Just wanted to confirm.

Reply:
Not sure, I can benchmark this

Reply:
it's about 3% faster to save the char rather than retrieving the char from the string each time according to my benchmarking